### PR TITLE
Add a simple filtering search endpoint to RQ

### PR DESCRIPTION
**Pull request by kumquatexpress**
Searches within a single queue for any messages that match a given set of parameters (url-encoded).
For params that don't exist in a message, will return empty.

### CODE REVIEW

**Comment by kumquatexpress** [comment]
Hmm, looking it at it doesn't ever return a non-enumerable, so that check can go away. I can adjust the search to be recursive over the JSON in params, basically a fold so that if all key-value matches are found within the object then it gets returned, if you think that would be a good idea.

**Comment by sodabrew** [comment]
The JSON PARAMs are only a convention, so they are just strings as far as RQ is concerned. And the convention in some cases is to prefix the PARAM with "json:" – like `RQ_PARAM1=json:{"foo":"bar"}`.

I would counter-propose that the search terms should be substring matches by default.

**Comment by kumquatexpress** [comment]
In a substring match, would the query then come in as `param1: "anything"` and be a match if "anything" exists anywhere in `param1`? It seems to me to be too generic, though there could be an option to do exact matching.

**Comment by sodabrew** [comment]
it's a slippery slope towards a query language ;) ... mainly I want any simple language now that does what you need leaves space for future query language options.

**Comment by kumquatexpress** [comment]
@sodabrew It was a struggle to figure out how to make tests work, but the current version passes. It's just ugly since they're all in the same method.

**Comment by drudru** [comment]
@sodabrew - when can 1.8.7 be dropped?

**Comment by sodabrew** [comment]
We're still in the lifecycle of Ubuntu Precise LTS, which has /usr/bin/ruby -> 1.8.7, so no time soon. I'll take a look at what might be breaking 1.8.7 for this PR, I'm excited for @kumquatexpress contribution!

**Comment by kumquatexpress** [comment]
It might be just on the test side, since the tests are the only thing that use Net::Http to consume the endpoint.

**Comment by drudru** [comment]
@sodabrew - Moving off of 1.8.7  is definitely something engineering should look at. It would get my vote :-)

**Comment by kumquatexpress** [comment]
@sodabrew @drudru hey guys, want to merge this in? It doesn't have to be deployed but at least it'll be one step closer.

**Comment by sodabrew** [comment]
Yeah, sorry - had a few fires earlier this week. I'm looking at it today.

**Comment by sodabrew** [comment]
Could you rebase once more?

**Comment by kumquatexpress** [comment]
Rebase complete

File: code/main.rb [hunk 1]
```diff
@@ -192,6 +192,46 @@ def flash(type, msg)
       queuemgr.queues.to_json
     end
 
+    get '/search' do
+      begin
+        qc = get_queueclient(params[:name])
+      rescue
+        throw :halt, [404, "404 - Queue not found"]
+      end
```
Comment (sodabrew) [inline code/main.rb]: You can remove this now, I have a `qc = ...` in the `before` block.

**Comment by sodabrew** [comment]
Thanks for rebasing! Please update to use the `qc` that's already available, and then I'll merge. I also want to make a UI for this soon-ish.

**Comment by kumquatexpress** [comment]
Wow, I just realized I've been rebasing/pushing to this branch in gitlab instead of here. Should be all up to date now

**Pull request merged**